doc: add docstrings to classes

woile opened the pull request.

woile commented:
For building my own framework, I understand that `Marker` is a key component of this library. Where the rest (`Container`, `Executor`, etc) would end up being mostly boilerplate, is this correct?

adriangb commented on di/executors/_sync.py:
```diff
@@ -2,6 +2,8 @@
 
 
 class SyncExecutor(SupportsSyncExecutor):
+    """Execute sync functions with the solved dependencies."""
+
```
```suggestion
    """An executor that executes only sync dependencies.
    
    Dependencies are executed sequentially.
    If any async dependencies are encountered an exception will be raised.
    If there are no async dependencies, this will be faster than using `AsyncExecutor` because there is no event loop overhead.
    """

```

woile commented on di/container/_container.py:
```diff
@@ -25,6 +25,14 @@
+
+    Usually hooked to the `App` in the framework.
+
+    Should happen once, maybe at app startup, because solving
+    dependencies can be slow.
```
Then it's not clear to me what should happen once. Or what should be the "initial" recommendation?
To have one container, or to have many containers that are solved once?

adriangb replied:
The recommendation is to have one container per "app" (using the term loosely) and call `solve()` for each dependency you want to solve (say each endpoint). The point we want to get across is that you should _not_ call `Container.solve()` every time you want to execute a dependency, you should call it beforehand. `Container.solve()` is slow.

woile replied:
How is xpresso doing it? I was thinking of adding an example to Container, to give users a rough idea.

> For example, the xpresso framework has one container per `Operation` (path + method)

adriangb replied:
Xpresso has 1 container per App and calls `Container.solve()` once per endpoint / lifespan:
- https://github.com/adriangb/xpresso/blob/main/xpresso/applications.py#L92
- https://github.com/adriangb/xpresso/blob/main/xpresso/routing/operation.py#L162-L169

adriangb replied:
I think something like:

> Generally you will want one Container per application. There is not performance advantage to re-using a container, the only reason to do so is to share binds. For each "thing" you want to wire with `di` and execute you'll want to call `Container.solve()` exactly once and then keep a reference to the returned SolvedDependant to pass to `Container.execute`. Solving is very expensive so avoid doing it in a hot loop.

woile replied:
updated

woile commented:
Shall I squash the commits?

I was wondering if we could refactor the [in-depth](https://www.adriangb.com/di/0.70.1/examples/#in-depth-example) example in a way that shows how to use it with different Requests and to showcase the cached solver. I'm having a hard time doing it. Thoughts?

adriangb commented:
> Shall I squash the commits?

No need, we'll do a squash merge at the end of the PR



> I was wondering if we could refactor the [in-depth](https://www.adriangb.com/di/0.70.1/examples/#in-depth-example) example in a way that shows how to use it with different Requests and to showcase the cached solver. I'm having a hard time doing it. Thoughts?

I think we could just add a for loop [here](https://github.com/adriangb/di/blob/main/docs_src/web_framework.py#L16) and say something like "for each incoming request execute the `SolvedDependant`. Is that what you were getting at?

woile commented:
Something more tangible like:

```py
web_framework = WebFramework()
web_framework.run(request1)
web_framework.run(request2)
```

Or with a for loop. But the idea is that it resembles more to the `App` model to me, similar to other frameworks.

adriangb commented:
> Something more tangible like:
> 
> ```python
> web_framework = WebFramework()
> web_framework.run(request1)
> web_framework.run(request2)
> ```
> 
> Or with a for loop. But the idea is that it resembles more to the `App` model to me, similar to other frameworks.

#84

adriangb commented:
Can you fix the linting errors?

adriangb commented:
Looking great! Let me know when you're done and I'll merge 😄

woile commented:
Done! Updated the Container with your comment, and removed the example for it which was no longer needed.

Thanks to you for the guidance. This library is saving me a lot of time and teaching me a lot as well 🤣 I was not aware of PEP593 and it's exactly what I'm trying to do now 😅

adriangb commented on di/executors/_sync.py:
```diff
@@ -2,6 +2,13 @@
+    """An executor that executes only sync dependencies.
+
+    Dependencies are executed sequentially.
+    If any async dependencies are encountered an exception will be raised.
```
```suggestion
    If any async dependencies are encountered a RuntimeError will be raised.
```
